idaes/core/base/costing_base.py

```diff
@@ -345,22 +345,33 @@ def register_flow_type(self, flow_type, cost):
         self.flow_types.add(flow_type)
 
         name = f"{flow_type}_cost"
-        current_component = self.component(name)
+        current_component = getattr(self, name, None)
         if current_component is not None:
             if current_component is not cost:
                 raise RuntimeError(
                     f"Component {name} already exists on {self} but is not {cost}."
                 )
             # now self.{flow_type}_cost is cost, so just use it
         else:
-            # Create a Var to hold the cost
-            # Units will be different between flows, so have to use scalar Vars
-            fvar = pyo.Var(
-                units=pyo.units.get_units(cost),
-                doc=f"Cost parameter for {flow_type} flow",
-            )
-            self.add_component(name, fvar)
-            fvar.fix(cost)
+            if is_constant_up_to_units(cost):
+                # Create a Var to hold the cost
+                # Units will be different between flows, so have to use scalar Vars
+                fvar = pyo.Var(
+                    units=pyo.units.get_units(cost),
+                    doc=f"Cost parameter for {flow_type} flow",
+                )
+                self.add_component(name, fvar)
+                fvar.fix(cost)
+            else:
+                add_object_reference(self, name, cost)
```

Comment:
I wonder if the is the best way to do this. Notably, this allows users to assign unnamed `expressions` for the cost but I wonder if it would be better to insist on named `Expressions` instead? This would save us from needing to do the `is_constant_unit_to_units` check, and might also be more robust to future changes.

I.e instead of allowing `costing.defined_flows["NaOCl"] = blk.cost / blk.purity` require:

```
blk.NaOCl_cost = pyo.Expression(expr=blk.cost / blk.purity)
costing.defined_flows["NaOCl"] = blk.NaOCl_cost
```

Reply:
We could force the user to prove named `Expressions`, or create one ourselves here.

I'm not sure that sidesteps the need for the `is_constant_up_to_units` check though. Wouldn't we still want to catch the case when we're evaluating the `value` of a non-constant `expression`?

Reply:
I would probably just do:

```
if isinstance(cost, [Var, Param, Expression]):
   add_object_reference(self, name, cost)
else:
   # Assume it is a constant and make a Var
```

Maybe add an `elif` to check for an unnamed `expressions` and raise a `TypeError`.

Reply:
> Maybe add an `elif` to check for an unnamed `expressions` and raise a `TypeError`.

Ah, that's exactly the rub though -- `0.2 * pyunits.J` (the original intended usage here) *is* and unnamed `expression`. Hence the gymnastics for evaluating unnamed `expression`s.

Reply:
Ah.... I see. Maybe we would still need the `is_constant_up_to_units`.

I am hoping @jsiirola might have some time to take a quick look at this and provide some thoughts.

Reply:
I am not sure why `register_flow_type` doesn't just put whatever the user handed it into an `Expression` and attach that to the block.  Does it matter if the cost is variable?  (I don't think so)

Reply:
@jsiirola If the user were to pass `0.2*units` as the argument however, this would make the cost a constant with no way to change it without changing the `expression` itself. The original API took this and created a `Var` so that users could change this latter for sensitivity studies or parameter estimation (or taking an existing costing package and changing values to suit a new use case).

Reply:
...but being able to change the `Expression` *after* constructing expressions is exactly what `Expression` components are for.  They are mutable named placeholders (or references, depending on how you want to look at it) in expressions -- and are the *only* mutable things in the expression system.

Reply:
That wouldn't work in a parameter estimation case however for example. Yes, the user could just have provided a `Var` upfront, but that would mean the current API no longer works (and I really don;t want to deprecated anything right now with IDAES 2.0 still not quite done).